### PR TITLE
Update SetContractSet to UpdateContractSet

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
Closes https://github.com/SiaFoundation/renterd/issues/1497

### CODE REVIEW

File: autopilot/contractor/contractor.go [hunk 1]
```diff
@@ -1343,7 +1343,17 @@ func performContractMaintenance(ctx *mCtx, alerter alerts.Alerter, bus Bus, chur
 	for _, contract := range newSet {
 		newSetIDs = append(newSetIDs, contract.ID)
 	}
-	if err := bus.SetContractSet(ctx, ctx.ContractSet(), newSetIDs); err != nil {
+	inNewSet := make(map[types.FileContractID]struct{})
```
Comment (peterjan) [inline autopilot/contractor/contractor.go:1346]: This is more correct I think:

```go
// added contracts are contracts not in the old set
var toAdd []types.FileContractID
inOldSet := make(map[types.FileContractID]struct{})
for _, c := range newSet {
    if _, ok := inOldSet[c.ID]; !ok {
        toAdd = append(toAdd, c.ID)
    }
    delete(toAdd, c.ID) // ensure we're left with contracts to remove
}

// removed contracts are old contracts not present in the new set
var toRemove []types.FileContractID
for id := range inOldSet {
    toRemove = append(toRemove, id)
}
```

File: bus/client/contracts.go [hunk 2]
```diff
@@ -193,8 +193,11 @@ func (c *Client) ReleaseContract(ctx context.Context, contractID types.FileContr
 	return
 }
 
-// SetContractSet adds the given contracts to the given set.
-func (c *Client) SetContractSet(ctx context.Context, set string, contracts []types.FileContractID) (err error) {
-	err = c.c.WithContext(ctx).PUT(fmt.Sprintf("/contracts/set/%s", set), contracts)
+// UpdateContractSet adds the given contracts to the given set.
```
Comment (peterjan) [inline bus/client/contracts.go:196]: It also removes contracts from the set.

File: bus/routes.go [hunk 3]
```diff
@@ -801,22 +801,23 @@ func (b *Bus) contractsSetsHandlerGET(jc jape.Context) {
 }
 
 func (b *Bus) contractsSetHandlerPUT(jc jape.Context) {
-	var contractIds []types.FileContractID
+	var req api.ContractSetUpdateRequest
 	if set := jc.PathParam("set"); set == "" {
 		jc.Error(errors.New("path parameter 'set' can not be empty"), http.StatusBadRequest)
```
Comment (peterjan) [inline bus/routes.go:806]: I'd also return a `400` if both `toAdd` and `toRemove` are `nil`

File: stores/sql/database.go [hunk 4]
```diff
@@ -317,9 +317,10 @@ type (
 		// substring.
 		SearchObjects(ctx context.Context, bucket, substring string, offset, limit int) ([]api.ObjectMetadata, error)
 
-		// SetContractSet creates the contract set with the given name and
-		// associates it with the provided contract IDs.
-		SetContractSet(ctx context.Context, name string, contractIds []types.FileContractID) error
+		// UpdateContractSet adds/removes the provided contract ids to/from
+		// the contract set. The contract set is created in the process if
+		// it doesn't eist already.
```
Comment (peterjan) [inline stores/sql/database.go:322]:
```suggestion
		// it doesn't exist already.
```

File: stores/sql/mysql/main.go [hunk 5]
```diff
@@ -862,6 +811,62 @@ func (tx *MainDatabaseTx) UpdateBucketPolicy(ctx context.Context, bucket string,
 	return ssql.UpdateBucketPolicy(ctx, tx, bucket, bp)
 }
 
+func (tx *MainDatabaseTx) UpdateContractSet(ctx context.Context, name string, toAdd, toRemove []types.FileContractID) error {
+	res, err := tx.Exec(ctx, "INSERT INTO contract_sets (name) VALUES (?) ON DUPLICATE KEY UPDATE id = last_insert_id(id)", name)
+	if err != nil {
+		return fmt.Errorf("failed to insert contract set: %w", err)
+	}
+
+	// if no changes are needed, return after creating the set
```
Comment (peterjan) [inline stores/sql/mysql/main.go:820]: Hm, still think a 400 is appropriate if both are empty.

Comment (ChrisSchinnerl) [reply stores/sql/mysql/main.go:820]: I think it's nice to have a way of creating a contract set without contracts. Even if it's just for some UI that let's you create an empty set before switching tabs and selecting contracts to the set.

**Review by peterjan — APPROVED** [review]

**Pull request merged**